From patchwork Sat Mar 16 09:16:21 2019
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Masanori Misono <m.misono760@gmail.com>
X-Patchwork-Id: 10855791
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 6C9DC139A
	for <patchwork-kvm@patchwork.kernel.org>;
 Sat, 16 Mar 2019 09:16:53 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 52CC62A783
	for <patchwork-kvm@patchwork.kernel.org>;
 Sat, 16 Mar 2019 09:16:53 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 455BF2A823; Sat, 16 Mar 2019 09:16:53 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-8.0 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,FREEMAIL_FROM,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI
	autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id E3CCF2A783
	for <patchwork-kvm@patchwork.kernel.org>;
 Sat, 16 Mar 2019 09:16:52 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1726649AbfCPJQv (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Sat, 16 Mar 2019 05:16:51 -0400
Received: from mail-pg1-f196.google.com ([209.85.215.196]:45088 "EHLO
        mail-pg1-f196.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1726571AbfCPJQv (ORCPT <rfc822;kvm@vger.kernel.org>);
        Sat, 16 Mar 2019 05:16:51 -0400
Received: by mail-pg1-f196.google.com with SMTP id 125so8058024pgc.12
        for <kvm@vger.kernel.org>; Sat, 16 Mar 2019 02:16:50 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20161025;
        h=from:to:cc:subject:date:message-id;
        bh=F6iKhVZd9X3guzPPdRAEawLyLnCgfSohWgf0YVhKXOI=;
        b=bna04f5X7kA0JECropX9vsaOlQ/+82GfU6gnZj/Nje0BqRkJi4NAihBhtZo8PO+OiP
         j4NBAvTeXyZK31lBuxP0G+3cqGwauxZ2RB+NueavddRH1HRfvRvdQHwfT81hKqpKSZfW
         lT8JESSBNVNIiqAtsbM/q9bai0I8T7OdstTpgoaNOzRA+YGNcabPFHcKKa1yAH349xKe
         UThOaIjnAea6QXnzHNKFg2kZ0DPzu6ffF8shykYtufQUfyKKwmh8shAiih40yvtVMUUf
         W31vMuFoIf7XMb5ld1WVy4TmUJYHKhdn8IiBqKJBUvB2IJGRd+hr5eHQNza4LrwyxEUj
         R9Tg==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:from:to:cc:subject:date:message-id;
        bh=F6iKhVZd9X3guzPPdRAEawLyLnCgfSohWgf0YVhKXOI=;
        b=nn9mJa+c+TbJ1F2J5B3edp3uhKo/DTEa8bITONB9Ux4aWuPnk4O6pocTgIvQjHwNJ4
         BXlHkfhuUGbTJNFo8CzZHLUNJGcL3GY2QGnCDKxEn2GdEod01+1BP+xAm53C78RvoiLb
         w4dKchGH9CpZ0g+QLAEps63DMzzY12J7H9ZlMDxW76EK9F9QVOsyNZbmfi2oDo7XuUXQ
         aCSxHkhwhAPzvk6+cTFjSJoIoxHL1m+e9hrMHSr/THbFXzK6vWOz358I6ch8OF9lBRB/
         mDX7GdWZe2QUXDjaAMe1xq12SOUbCnWR7T0EKbiH/sGBdBd6l5OiDLuXWSdQh2hFNs/a
         4Q0w==
X-Gm-Message-State: APjAAAUU+7PhCfX85hyx7NnLBuLRNUyJ/mG5GsUUd2m4zzTeBJTMY/8L
        Q/yb+np+8uxvARFkEvvj50zWZUnG
X-Google-Smtp-Source: 
 APXvYqxXdI8c3ddGb9/05hKzFXhFA7FnMQHaGsHXGKKzPJlpPBuuWbsxLjk/g3Xf5UiLbt/+IVghyQ==
X-Received: by 2002:aa7:8a59:: with SMTP id n25mr8665049pfa.251.1552727809412;
        Sat, 16 Mar 2019 02:16:49 -0700 (PDT)
Received: from localhost.localdomain ([133.11.45.40])
        by smtp.gmail.com with ESMTPSA id
 n6sm12993439pgh.79.2019.03.16.02.16.48
        (version=TLS1_2 cipher=ECDHE-RSA-AES128-GCM-SHA256 bits=128/128);
        Sat, 16 Mar 2019 02:16:48 -0700 (PDT)
From: Masanori Misono <m.misono760@gmail.com>
To: kvm@vger.kernel.org
Cc: Masanori Misono <m.misono760@gmail.com>
Subject: [PATCH] KVM: x86: check KVM availability when setting up pv tlb flush
Date: Sat, 16 Mar 2019 09:16:21 +0000
Message-Id: <20190316091621.2198-1-m.misono760@gmail.com>
X-Mailer: git-send-email 2.17.1
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

kvm_setup_pv_tlb_flush() does not check if kvm_para_available() is true
unlike kvm_guest_init(). Therefore it may proceed even if
kvm_para_available() is 0 (e.g., when booting the VM with `qemu -cpu
host,kvm=off`). In this case, the pr_info message is confusing.
This fixed it.

Signed-off-by: Masanori Misono <m.misono760@gmail.com>
---
 arch/x86/kernel/kvm.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/arch/x86/kernel/kvm.c b/arch/x86/kernel/kvm.c
index 5c93a65ee1e5..869ac19c52cf 100644
--- a/arch/x86/kernel/kvm.c
+++ b/arch/x86/kernel/kvm.c
@@ -69,6 +69,7 @@ early_param("no-steal-acc", parse_no_stealacc);
 static DEFINE_PER_CPU_DECRYPTED(struct kvm_vcpu_pv_apf_data, apf_reason) __aligned(64);
 static DEFINE_PER_CPU_DECRYPTED(struct kvm_steal_time, steal_time) __aligned(64);
 static int has_steal_clock = 0;
+static int has_pv_tlb_flush = 0;
 
 /*
  * No need for any "IO delay" on KVM
@@ -641,6 +642,7 @@ static void __init kvm_guest_init(void)
 	if (kvm_para_has_feature(KVM_FEATURE_PV_TLB_FLUSH) &&
 	    !kvm_para_has_hint(KVM_HINTS_REALTIME) &&
 	    kvm_para_has_feature(KVM_FEATURE_STEAL_TIME)) {
+		has_pv_tlb_flush = 1;
 		pv_ops.mmu.flush_tlb_others = kvm_flush_tlb_others;
 		pv_ops.mmu.tlb_remove_table = tlb_remove_table;
 	}
@@ -748,9 +750,7 @@ static __init int kvm_setup_pv_tlb_flush(void)
 {
 	int cpu;
 
-	if (kvm_para_has_feature(KVM_FEATURE_PV_TLB_FLUSH) &&
-	    !kvm_para_has_hint(KVM_HINTS_REALTIME) &&
-	    kvm_para_has_feature(KVM_FEATURE_STEAL_TIME)) {
+	if (has_pv_tlb_flush){
 		for_each_possible_cpu(cpu) {
 			zalloc_cpumask_var_node(per_cpu_ptr(&__pv_tlb_mask, cpu),
 				GFP_KERNEL, cpu_to_node(cpu));
